groups/csa/complexity_check.cpp

```diff
@@ -0,0 +1,143 @@
+#include <clang/ASTMatchers/ASTMatchers.h>
+#include <clang/ASTMatchers/ASTMatchFinder.h>
+#include <clang/Lex/Lexer.h>
+#include <clang/AST/Stmt.h>
+
+#include <csabase_analyser.h>
+#include <csabase_debug.h>
+#include <csabase_diagnostic_builder.h>
+#include <csabase_registercheck.h>
+#include <csabase_report.h>
+#include <csabase_util.h>
+#include <csabase_visitor.h>
+
+#include <iostream>
+#include <sstream>
+
+using namespace csabase;
+using namespace clang;
+using namespace clang::ast_matchers;
+
+static std::string const check_name("complexity-check");
+
+namespace {
+    
+struct data {
+    // Maximum complexity per method
+    int d_max_complexity;
+};
+
+
+struct report : Report<data> {
+
+    INHERIT_REPORT_CTOR(report, Report, data);
+
+    void calculateComplex(BoundNodes const & nodes);
+    void operator()();
+    void generateReport();
+
+    // Maps each function to its complexity value
+    std::map<FunctionDecl const *, int> complexDict;
+    // Total file complexity
+    int complexity = 0;
+};
+
+
+// Matches every if,switch,while,for,try,catch inside a method declaration
+static const internal::DynTypedMatcher& dyn_matchComplexity() 
+{
+    static const internal::DynTypedMatcher& matcher =
+    findAll(functionDecl(hasDescendant(stmt(eachOf(
+        findAll(stmt(ifStmt().bind("if"))), 
+        findAll(stmt(switchCase().bind("switch"))),
+        findAll(stmt(whileStmt().bind("while"))),
+        findAll(stmt(forStmt().bind("for"))),
+        findAll(stmt(tryStmt().bind("try"))),
+        findAll(stmt(catchStmt().bind("catch"))),
+        findAll(stmt(conditionalOperator().bind("?operator"))),
+        findAll(stmt(ifStmt(hasCondition(findAll(binaryOperator(hasOperatorName("&&")).bind("condOp")))))),
+        findAll(stmt(ifStmt(hasCondition(findAll(binaryOperator(hasOperatorName("||")).bind("condOp"))))))
+        )))).bind("methodDecl"));
```

Comment:
Call it "function" instead of Method, a method is a function but a function may not be a method.